### PR TITLE
p7zip: 17.04 -> 17.05

### LINKED ISSUE #216766: p7zip source is gone
## Issue description

p7zip source is gone

### Steps to reproduce

try to build p7zip.src without substituters

cc @7c6f434c @06kellyjac

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Description of changes

Updated to 17.05

* Moved back to p7zip-project release now it's being updated
* prePatch was not being used due to no patches being present
* Split outputs

Closes #216766

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/extractcode/7z.nix [hunk 1]
```diff
@@ -27,7 +27,7 @@ buildPythonPackage rec {
 
     rm 7z 7z.so
     ln -s ${p7zip}/bin/7z 7z
-    ln -s ${p7zip}/lib/p7zip/7z.so 7z.so
+    ln -s ${p7zip.lib}/lib/p7zip/7z.so 7z.so
```
Comment (7c6f434c) [inline pkgs/development/python-modules/extractcode/7z.nix:30]: Maybe getLib then?

Comment (06kellyjac) [reply pkgs/development/python-modules/extractcode/7z.nix:30]: Sure

File: pkgs/tools/archivers/p7zip/default.nix [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
-    rev    = "v${version}";
+    owner = "p7zip-project";
+    repo = "p7zip";
+    rev = "v${finalAttrs.version}";
     sha256 = {
-      free = "sha256-DrBuf2VPdcprHI6pMSmL7psm2ofOrUf0Oj0qwMjXzkk=";
-      unfree = "sha256-19F4hPV0nKVuFZNbOcXrcA1uW6Y3HQolaHVIYXGmh18=";
+      free = "sha256-4byb9bb3FxV9e3NnKHOwCBd8jumyzGr/SUYy6AoaY1Y=";
+      unfree = "sha256-SO1XtqGFnaXLIvGitZyJGra4O6cAbO2aiiCMfhlFpNA=";
     }.${if enableUnfree then "unfree" else "free"};
```
Comment (SuperSandro2000) [inline pkgs/tools/archivers/p7zip/default.nix:14]: This could be made simpler if only an if else.

Comment (06kellyjac) [reply pkgs/tools/archivers/p7zip/default.nix:14]: Very true but I would need to also change the update script which I really don't want to do ATM

If you do I'd be happy to add the change

File: pkgs/tools/archivers/p7zip/default.nix [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
-    owner  = "jinfeihan57";
-    repo   = pname;
-    rev    = "v${version}";
+    owner = "p7zip-project";
+    repo = "p7zip";
+    rev = "v${finalAttrs.version}";
     sha256 = {
-      free = "sha256-DrBuf2VPdcprHI6pMSmL7psm2ofOrUf0Oj0qwMjXzkk=";
-      unfree = "sha256-19F4hPV0nKVuFZNbOcXrcA1uW6Y3HQolaHVIYXGmh18=";
+      free = "sha256-4byb9bb3FxV9e3NnKHOwCBd8jumyzGr/SUYy6AoaY1Y=";
```
Comment (SuperSandro2000) [inline pkgs/tools/archivers/p7zip/default.nix:12]:
```suggestion
      free = "sha256-5r7M9BVcAryZNTkqJ/BfHnSSWov1PwoZhUnLBwEbJoA=";
```

File: pkgs/tools/archivers/p7zip/default.nix [hunk 4]
```diff
@@ -22,40 +22,33 @@ stdenv.mkDerivation rec {
     '';
   };
 
-  # Default makefile is full of impurities on Darwin. The patch doesn't hurt Linux so I'm leaving it unconditional
-  postPatch = ''
```
Comment (jtojnar) [inline pkgs/tools/archivers/p7zip/default.nix:26]: > prePatch was not being used due to no patches being present

There is only `postPatch` and the `patchPhase` runs the hooks unless overridden or disabled even when there are no patches:

https://github.com/NixOS/nixpkgs/blob/287805fcb77f914fb53add71eabfe274dd6e9f55/pkgs/stdenv/generic/setup.sh#L1200

File: pkgs/tools/archivers/p7zip/default.nix [hunk 5]
```diff
@@ -22,40 +22,33 @@ stdenv.mkDerivation rec {
...
-  setupHook = ./setup-hook.sh;
+  makeFlags = [
+    "DEST_BIN=${placeholder "out"}/bin"
+    "DEST_SHARE=${placeholder "lib"}/lib/p7zip"
+    "DEST_MAN=${placeholder "man"}/share/man"
+    "DEST_SHARE_DOC=${placeholder "doc"}/share/doc/p7zip"
+  ];
 
-  env.NIX_CFLAGS_COMPILE = lib.optionalString stdenv.cc.isClang "-Wno-error=c++11-narrowing";
+  outputs = [ "out" "lib" "doc" "man" ];
```
Comment (jtojnar) [inline pkgs/tools/archivers/p7zip/default.nix:45]: Outputs typically go between `version` and `src` https://github.com/jtojnar/nixpkgs-hammering/blob/243b81c687aac33d6716957c0cd2235c81631044/explanations/attribute-ordering.md

Comment (06kellyjac) [reply pkgs/tools/archivers/p7zip/default.nix:45]: > This is nothing but a personal preference but [growing stronger](https://discourse.nixos.org/t/document-attribute-ordering-in-package-expressions/4887).

With the link only having yourself mention outputs.

Respectfully I'd like to keep it here near where install related stuff tends to go

File: pkgs/tools/archivers/p7zip/default.nix (hunk 4 above)
Comment (06kellyjac) [reply pkgs/tools/archivers/p7zip/default.nix:26]: not sure what I did here :upside_down_face: reverted

**Pull request merged**